### PR TITLE
Move `duplicateTemplatePart` action to the `@wordpress/fields` package

### PR DESCRIPTION
**Pull request by gigitux**
## What?

This PR migrates the `duplicateTemplatePart` action to the `@wordpress/fields` package. The primary issue with this action is its dependency on the `CreateTemplatePartModalContents` component, which resides in the `@wordpress/editor` package. Due to this, it can't be directly used within the `@wordpress/fields` package.

For the purpose of the `@wordpress/fields` package work, I have temporarily moved the component to this package. However, I'm not certain if this is the best long-term solution. Ideally, this component (and potentially others) should reside in a dedicated package, such as `@wordpress/template`, which would be more appropriate for components related to template and template-parts. A similar package for the patterns already exists: https://github.com/WordPress/gutenberg/blob/d6fcf4aea4aff17cf0910862bf10090e20ba2ab0/packages/patterns/src

Please review and provide feedback on whether this approach is suitable or if there are alternative recommendations for handling these dependencies. Thanks! 🙏 cc @youknowriad @oandregal 

## Testing Instructions

Ensure that `Custom Dataviews` is enabled.

1. Visit the Site Editor.
2. Visit the patterns view.
3. Click on header
4. Duplicate a header template part.
5. Ensure that the duplicated template part is created.

### CODE REVIEW

File: packages/fields/src/components/create-template-part-modal/index.tsx [hunk 1]
```diff
@@ -152,10 +208,14 @@ export function CreateTemplatePartModalContents( {
 						label={ __( 'Area' ) }
 						className="editor-create-template-part-modal__area-radio-group"
```
Comment (gigitux) [inline packages/fields/src/components/create-template-part-modal/index.tsx]: Should I update these classes even if it is a temporary solution?

**Comment by oandregal** [comment]
@gigitux what's the status on this one? I saw https://github.com/WordPress/gutenberg/pull/65289 has been approved but not merged yet.

**Comment by gigitux** [comment]
> @gigitux what's the status on this one? I saw #65289 has been approved but not merged yet.

@oandregal I rebased this PR after #65289 has been merged. Let me know if you agree with the current method, or if you believe a dedicated package should be created for these template-related components.

P.S: As reminder, we are in cooldown, so I'm not strictly focused on this project. I may be slow (I will do my best) on checking your feedback 🙇

**Comment by gigitux** [comment]
I rebased this PR and it is ready to be reviewed. My main concern is this one:

> For the purpose of the @wordpress/fields package work, I have temporarily moved the component to this package. However, I'm not certain if this is the best long-term solution. Ideally, this component (and potentially others) should reside in a dedicated package, such as @wordpress/template, which would be more appropriate for components related to template and template-parts. A similar package for the patterns already exists: https://github.com/WordPress/gutenberg/blob/d6fcf4aea4aff17cf0910862bf10090e20ba2ab0/packages/patterns/src

File: packages/editor/src/components/template-part-menu-items/convert-to-template-part.js [hunk 2]
```diff
@@ -13,7 +13,7 @@ import { symbolFilled } from '@wordpress/icons';
 /**
  * Internal dependencies
  */
-import CreateTemplatePartModal from '../create-template-part-modal';
+import { CreateTemplatePartModal } from '@wordpress/fields';
```
Comment (youknowriad) [inline packages/editor/src/components/template-part-menu-items/convert-to-template-part.js:16]: Yes, I agree this is weird, but also not sure what's the best place for these wp-core-data aware components. That said I can live with it being in fields for now.

Comment (youknowriad) [reply packages/editor/src/components/template-part-menu-items/convert-to-template-part.js:16]: Maybe we should add a comment explaining the choice.

File: packages/fields/src/constants.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+export const TEMPLATE_POST_TYPE = 'wp_template';
+export const TEMPLATE_PART_POST_TYPE = 'wp_template_part';
+export const TEMPLATE_ORIGINS = {
+	custom: 'custom',
+	theme: 'theme',
+	plugin: 'plugin',
+};
+
+export const TEMPLATE_PART_AREA_DEFAULT_CATEGORY = 'uncategorized';
```
Comment (youknowriad) [inline packages/fields/src/constants.ts]: Personally I hate these constants file and would rather mostly remove them (especially since most of these types are "typed" now.

File: packages/fields/src/components/create-template-part-modal/index.tsx [hunk 4]
```diff
@@ -79,11 +130,14 @@ export function CreateTemplatePartModalContents( {
...
-	const templatePartAreas = useSelect(
-		( select ) =>
-			select( editorStore ).__experimentalGetDefaultTemplatePartAreas(),
+	const settings = useSelect(
+		// @ts-ignore
+		( select ) => select( blockEditorStore ).getSettings(),
 		[]
 	);
+
+	const defaultTemplatePartAreas = getDefaultTemplatePartAreas( settings );
```
Comment (youknowriad) [inline packages/fields/src/components/create-template-part-modal/index.tsx]: Personally, I think this is the wrong solution to retrieving the default template areas. There might not be a rendered block-editor when using the fields, so I don't think a dependency to "block-editor" is the right approach.

The main issue here is that `defaultTemplateTypes` and `defaultTemplatePartAreas ` are passed from the server to the client through the editor settings initialization. I think this is the wrong approach. The right approach would be to move these to `getEntityRecord( 'root', '__unstableBase' )` endpoint and add them in the server to this endpoint. I believe we did things like that in the past for things like `site_icon_url` ...  Do you think this is something we can try in its own PR. I'm a bit concerned that the current field won't work as expected in all contexts.

File: packages/editor/src/components/template-part-menu-items/convert-to-template-part.js (hunk 2 above)
Comment (gigitux) [reply packages/editor/src/components/template-part-menu-items/convert-to-template-part.js:16]: With 47da44097c17f402a430b1a0a1a606ced40053dc I added some documentation.

File: packages/fields/src/constants.ts (hunk 3 above)
Comment (gigitux) [reply packages/fields/src/constants.ts]: Removed with 64423cf8ff147ab7d19b7acfee209f2b5fbe8f97

File: packages/fields/src/components/create-template-part-modal/index.tsx (hunk 4 above)
Comment (gigitux) [reply packages/fields/src/components/create-template-part-modal/index.tsx]: > Do you think this is something we can try in its own PR. I'm a bit concerned that the current field won't work as expected in all contexts.

I'm happy to work on it with a dedicated PR! 💪

File: packages/fields/src/components/create-template-part-modal/index.tsx (hunk 1 above)
Comment (youknowriad) [reply packages/fields/src/components/create-template-part-modal/index.tsx]: Yeah I think we should update them ideally `fields` instead of `editor` basically.

Comment (gigitux) [reply packages/fields/src/components/create-template-part-modal/index.tsx]: I updated it with 199c313b51f724a6f26bfc186e3a0e579e369d27

File: packages/fields/src/components/create-template-part-modal/index.tsx (hunk 4 above)
Comment (gigitux) [reply packages/fields/src/components/create-template-part-modal/index.tsx]: I created #66459. Let me know if you agree with the approach that I took!

**Comment by gigitux** [comment]
This PR is ready to be reviewed! cc @oandregal @youknowriad @louwie17

File: packages/fields/README.md [hunk 5]
```diff
@@ -22,6 +22,30 @@ Author field for BasePost.
...
+_Parameters_
+
+-   _props_ `Object`: - The component props.
+-   _props.defaultArea_ `[string]`: - The default area for the template part.
+-   _props.blocks_ `[Array]`: - The blocks to be included in the template part.
+-   _props.confirmLabel_ `[string]`: - The label for the confirm button.
+-   _props.closeModal_ `Function`: - Function to close the modal.
+-   _props.onCreate_ `Function`: - Function to call when the template part is successfully created.
+-   _props.onError_ `[Function]`: - Function to call when there is an error creating the template part.
+-   _props.defaultTitle_ `[string]`: - The default title for the template part.
```
Comment (youknowriad) [inline packages/fields/README.md]: Do we need both components to be exposed?

File: packages/fields/README.md [hunk 6]
```diff
@@ -42,6 +66,14 @@ Undocumented declaration.
 
 Undocumented declaration.
 
+### duplicateTemplatePart
```
Comment (youknowriad) [inline packages/fields/README.md:58]: Do we need this to be a public API here?

Comment (youknowriad) [reply packages/fields/README.md:58]: Are we using it elsewhere?

File: packages/fields/package.json [hunk 7]
```diff
@@ -35,6 +35,7 @@
 		"@babel/runtime": "7.25.7",
 		"@wordpress/api-fetch": "*",
 		"@wordpress/blob": "*",
+		"@wordpress/block-editor": "*",
```
Comment (youknowriad) [inline packages/fields/package.json]: Why is this needed?

File: packages/fields/README.md (hunk 6 above)
Comment (gigitux) [reply packages/fields/README.md:58]: We are using only [here](https://github.com/WordPress/gutenberg/blob/084383c03305e6cec722ad8c48eedef874fa7346/packages/editor/src/dataviews/store/private-actions.ts#L142-L146). I'm happy to make a “private” API.

File: packages/fields/README.md (hunk 5 above)
Comment (gigitux) [reply packages/fields/README.md]: Fixed with 4583cf82c0f432c0b72c154f7008ecd3041767af

File: packages/fields/package.json (hunk 7 above)
Comment (gigitux) [reply packages/fields/package.json]: Fixed with e0f518e4912a0cd7ebaf2af46e8e93cc2ac29da5.

File: packages/fields/README.md (hunk 6 above)
Comment (youknowriad) [reply packages/fields/README.md:58]: ah no sorry, it's an action, let's keep it. I thought it was some utility or something.

File: packages/fields/package.json (hunk 7 above)
Comment (youknowriad) [reply packages/fields/package.json]: Is this commit not in the branch? For some reason GitHub is showing me the dependency addition still in the diff

File: packages/fields/src/actions/duplicate-template-part.tsx [hunk 8]
```diff
@@ -61,7 +63,7 @@ const duplicateTemplatePart: Action< TemplatePart > = {
 				onCreate={ onTemplatePartSuccess }
 				onError={ closeModal }
 				confirmLabel={ _x( 'Duplicate', 'action label' ) }
-				closeModal={ closeModal }
+				closeModal={ closeModal ?? ( () => void 0 ) }
```
Comment (youknowriad) [inline packages/fields/src/actions/duplicate-template-part.tsx]: Why `void 0` and not `{}`

File: packages/fields/src/index.ts [hunk 9]
```diff
@@ -1,3 +1,4 @@
 export * from './fields';
 export * from './actions';
-export type * from './types';
+export { default as CreateTemplatePartModal } from './components/create-template-part-modal';
+export * from './types';
```
Comment (youknowriad) [inline packages/fields/src/index.ts]: Do we need to export the types?

Comment (gigitux) [reply packages/fields/src/index.ts]: Currently, we are using some types: https://github.com/WordPress/gutenberg/pull/67024/files#diff-22dafac83020324ae032e8217a80eabc59ddc3eedcb59c62d7353f4e49739404R18 and https://github.com/gigitux/gutenberg/blob/02da400ff55249f6b822cbf0a0e13fdfca1bbc0b/packages/editor/src/dataviews/store/private-actions.ts#L13-L14

I'm going to export only the used one.

Comment (gigitux) [reply packages/fields/src/index.ts]: Fixed with b9fb6f3a679dc9b05f4b7f5f38ce56b997852b8d

File: packages/fields/src/actions/duplicate-template-part.tsx (hunk 8 above)
Comment (gigitux) [reply packages/fields/src/actions/duplicate-template-part.tsx]: Fixed with bb1f3668c2a43426e1899d74bd278989b787d024

File: packages/fields/package.json (hunk 7 above)
Comment (gigitux) [reply packages/fields/package.json]: It should be okay now!

File: packages/fields/src/index.ts (hunk 9 above)
Comment (youknowriad) [reply packages/fields/src/index.ts]: It's a bit unfortunate that they show up in the docs (not in a great way)

**Review by youknowriad — APPROVED** [review]
Thanks LGTM

**Comment by gigitux** [comment]
> Thanks LGTM

Thanks for your review! 🙇

File: packages/editor/src/components/template-part-menu-items/convert-to-template-part.js (hunk 2 above)
Comment (oandregal) [reply packages/editor/src/components/template-part-menu-items/convert-to-template-part.js:16]: I've looked where this modal is being used. It looks like there are two additional places other than the action:

| Patterns > Add new | Block Toolbar | 
| --- | --- |
| Screenshot 2024-12-04 at 11 57 19 | Screenshot 2024-12-04 at 11 58 56 |

The second case is a bit more difficult to absorb declaratively because it's very custom. However, it seems the first one ("add new") we should be able to absorb as part of the actions registration. Note how all entities have it. It's a bit complex in patterns due to how they are setup, but I presume we want the same design ("add new" button) for any new registered CPT:

| Patterns | Templates | Pages |
| --- | --- | --- |
| Screenshot 2024-12-04 at 11 57 19 | Screenshot 2024-12-04 at 12 02 12 | Screenshot 2024-12-04 at 12 02 03 |

We could look at declare some actions as "insertion", similarly to how some of them are "destructive", perhaps.

File: packages/fields/src/components/create-template-part-modal/index.tsx [hunk 10]
```diff
@@ -16,59 +15,107 @@ import {
...
 	modalTitle,
 	...restProps
-} ) {
+}: {
+	modalTitle: string;
+} & CreateTemplatePartModalContentsProps ) {
 	const defaultModalTitle = useSelect(
 		( select ) =>
-			select( coreStore ).getPostType( TEMPLATE_PART_POST_TYPE )?.labels
+			// @ts-ignore
```
Comment (ciampo) [inline packages/fields/src/components/create-template-part-modal/index.tsx:67]: Is this `// @ts-ignore` necessary?

In general, I usually recommend using `@ts-expect-error` instead of `@ts-ignore` ([this article](https://dev.to/maafaishal/ts-expect-error-over-ts-ignore-in-typescript-5140) explains the reason).

It's also usually a good idea to leave an inline comment explaining why we're ignoring/expecting a given TS error.

Comment (gigitux) [reply packages/fields/src/components/create-template-part-modal/index.tsx:67]: Nice catch! I overlooked it when I iterate on this PR. Fixed with https://github.com/WordPress/gutenberg/pull/67709
